src/xheartbeat_client.cpp

```diff
@@ -0,0 +1,68 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay, Martin Renou          *
+* Copyright (c) 2016, QuantStack                                           *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#include "xheartbeat_client.hpp"
+#include "xclient_zmq_impl.hpp"
+
+#include "xeus-zmq/xmiddleware.hpp"
+
+namespace xeus
+{
+
+    xheartbeat_client::xheartbeat_client(zmq::context_t& context,
+                                 const xeus::xconfiguration& config)
+        : m_heartbeat(context, zmq::socket_type::req)
+        , m_controller(context, zmq::socket_type::rep)
+    {
+        m_heartbeat.connect(get_end_point(config.m_transport, config.m_ip, config.m_hb_port));
+        init_socket(m_controller, get_controller_end_point("heartbeat"));
+    }
+
+    xheartbeat_client::~xheartbeat_client()
+    {
+    }
+
+    void xheartbeat_client::send_heartbeat_message()
+    {
+        zmq::message_t ping_msg("ping", 4);
+        m_heartbeat.send(ping_msg, zmq::send_flags::none);
+    }
+
+    bool xheartbeat_client::wait_for_answer(long timeout)
+    {
+        m_heartbeat.set(zmq::sockopt::linger, static_cast<int>(timeout));
+        zmq::message_t response;
+        return m_heartbeat.recv(response).has_value();
+    }
+
+    void xheartbeat_client::run(long timeout)
+    {
+        bool stop = false;
+        int retry_count = 0;
+        const int max_retries = 3;
+
+        while(!stop)
+        {
+            send_heartbeat_message();
+            if(!wait_for_answer(timeout))
+            {
+                if (retry_count < max_retries)
+                {
+                    ++retry_count;
+                } else {
+                    p_client_impl->notify_kernel_dead(true);
```

Comment:
I've registered a `kernel_status_listener` in the `xclient_zmq_impl` class. Does that work ?

Reply:
I think it would be better to register it in the `heartbeat_client` to avoid a circular dependency `heartbeat_client` <-> `client_zmq_impl`

Reply:
Noted